constellation-app/constellation
# Save screenshot as hash to resolve screenshot issue with the same name.

### Prerequisites


    for the "CI / Unit Tests") to complete first.

### Description of the Change

Addresses the issue of graphs with the same name having the same screenshot on the welcome page.  In this implementation the graph screenshot is saved as the MD5 hash of the path instead of the filename.  If a graph with the same name is saved in two different directories it will display the correct screenshot instead of the most recent graph with the same name.  It is also backward compatible with the graph name, if it can't find the hash of the recent graph it will look for the graph name, failing that it will use the default.  

![image](https://user-images.githubusercontent.com/63685177/138592387-d0263920-ce55-4e97-b648-cd0c5464f411.png)

### Alternate Designs

We could have included the path in the screenshot name  but this could have had some weird side effects when saving the file.  We could have also included just the parent directory in the name but we may have had the same issue with the same name and same parent folder even if it was less likely. 

### Why Should This Be In Core?

Bugfix addressing #1499.

### Benefits

It better aligns what the user is expecting when they open the welcome page and gives a better representation of the recent graphs.

### Possible Drawbacks

If the graph is moved in the file system without saving in constellation then the path will no longer match the hash in the screenshots directory and will display the default icon instead of the screenshot. 

There is a very very small chance that the md5 hash collides with a path that isn't a match.  

### Verification Process

- Create a new graph
- Add a few nodes and save the graph
- Create a new graph
- Create a sphere graph (or something else, so long as the graph is different to the first one)
- Save the graph in a different location but with the same name as the previous graph
- Close and open Constellation
- The screenshots for each graph on the welcome page with the same name should match the graph it is opening instead of the same screenshot for all the graphs that have the same name.

### Applicable Issues

#1499
